File: foodOrderingSystem/java-without-multithreading/services/FoodOrderingSystem.java
package services;
import entities.Order;
import entities.Restaurant;
import strategy.RestaurantSelectionStrategy;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

public class FoodOrderingSystem {
    private Map<String, Restaurant> restaurants;
    private Map<String, Order> orders;
    private static FoodOrderingSystem instance;

    private FoodOrderingSystem() {
        this.restaurants = new ConcurrentHashMap<>();
        this.orders = new ConcurrentHashMap<>();
    }

    public static synchronized FoodOrderingSystem getInstance() {
        if (instance == null) {
            instance = new FoodOrderingSystem();
        }
        return instance;
    }

    public void onboardRestaurant(String name, int maxOrders, double rating, Map<String, Double> menu) {
        Restaurant r = new Restaurant(name, maxOrders, rating);
        for(Map.Entry<String, Double> entry : menu.entrySet()) {
            r.updateMenu(entry.getKey(), entry.getValue());
        }
        restaurants.put(name, r);
    }

    public void updateMenu(String name, String action, String itemName, Double price) throws Exception {
        if (!restaurants.containsKey(name)) throw new Exception("Restaurant unmapped natively.");
        if (action.equalsIgnoreCase("add") || action.equalsIgnoreCase("update")) {
            restaurants.get(name).updateMenu(itemName, price);
        }
    }

    public void updateCapacity(String name, int capacity) throws Exception {
        if (!restaurants.containsKey(name)) throw new Exception("Restaurant computationally unmapped.");
        restaurants.get(name).updateCapacity(capacity);
    }

    public Order placeOrder(String orderId, String user, Map<String, Integer> items, RestaurantSelectionStrategy strategy) throws Exception {
        List<Restaurant> eligibleRestaurants = new ArrayList<>();
        
        // Validation boundary: filter exclusively restaurants physically serving all requested mapping items AND having safe capacity runtime limits!
        for (Restaurant r : restaurants.values()) {
            if (r.canFulfill(items)) {
                eligibleRestaurants.add(r);
            }
        }

        if (eligibleRestaurants.isEmpty()) {
            throw new Exception("Output: Order can't be fulfilled");
        }

        Optional<Restaurant> selectedOpt = strategy.selectRestaurant(eligibleRestaurants, items);
        if (!selectedOpt.isPresent()) {
            throw new Exception("Strategy isolated mathematically matching zero natively.");
        }

        Restaurant selected = selectedOpt.get();
        selected.incrementCurrentOrders();
        
        Order order = new Order(orderId, user, selected.getName(), items);
        orders.put(orderId, order);
        
        System.out.printf("Output: Order assigned to %s\n", selected.getName());
        return order;
    }

    public void updateOrderStatus(String orderId, enums.OrderStatus status) throws Exception {
        if (!orders.containsKey(orderId)) throw new Exception("Order strictly unmapped natively.");
        Order o = orders.get(orderId);
        
        if (o.getStatus() == enums.OrderStatus.COMPLETED) {
            throw new Exception("Order structurally already explicitly processed.");
        }

        if (status == enums.OrderStatus.COMPLETED) {
            o.setStatus(status);
            restaurants.get(o.getRestaurantName()).markOrderCompleted();
        }
    }
}

File: foodOrderingSystem/go-without-multithreading/services/food_ordering_system.go
package services

import (
	"app/foodOrderingSystem/go-without-multithreading/entities"
	"app/foodOrderingSystem/go-without-multithreading/enums"
	"app/foodOrderingSystem/go-without-multithreading/strategy"
	"fmt"
)

type FoodOrderingSystem struct {
	Restaurants map[string]*entities.Restaurant
	Orders      map[string]*entities.Order
}

var instance *FoodOrderingSystem

func GetInstance() *FoodOrderingSystem {
	if instance == nil {
		instance = &FoodOrderingSystem{
			Restaurants: make(map[string]*entities.Restaurant),
			Orders:      make(map[string]*entities.Order),
		}
	}
	return instance
}

func (s *FoodOrderingSystem) OnboardRestaurant(name string, maxOrders int, rating float64, menu map[string]float64) {
	r := entities.NewRestaurant(name, maxOrders, rating)
	for item, price := range menu {
		r.UpdateMenu(item, price)
	}
	s.Restaurants[name] = r
}

func (s *FoodOrderingSystem) UpdateMenu(name string, action string, itemName string, price float64) error {
	restaurant, exists := s.Restaurants[name]
	if !exists {
		return fmt.Errorf("restaurant structurally unmapped natively")
	}

	if action == "add" || action == "update" {
		restaurant.UpdateMenu(itemName, price)
	}
	return nil
}

func (s *FoodOrderingSystem) UpdateCapacity(name string, capacity int) error {
	restaurant, exists := s.Restaurants[name]
	if !exists {
		return fmt.Errorf("restaurant mathematically unmapped natively")
	}
	restaurant.UpdateCapacity(capacity)
	return nil
}

func (s *FoodOrderingSystem) PlaceOrder(orderId, user string, items map[string]int, strat strategy.RestaurantSelectionStrategy) (*entities.Order, error) {
	var eligibleRestaurants []*entities.Restaurant

	for _, r := range s.Restaurants {
		if r.CanFulfill(items) {
			eligibleRestaurants = append(eligibleRestaurants, r)
		}
	}

	if len(eligibleRestaurants) == 0 {
		return nil, fmt.Errorf("Output: Order can't be fulfilled")
	}

	selected := strat.SelectRestaurant(eligibleRestaurants, items)
	if selected == nil {
		return nil, fmt.Errorf("strategy failed mathematical match cleanly")
	}

	selected.IncrementCurrentOrders()

	order := &entities.Order{
		OrderID:        orderId,
		CustomerName:   user,
		RestaurantName: selected.Name,
		Items:          items,
		Status:         enums.Accepted,
	}
	s.Orders[orderId] = order

	fmt.Printf("Output: Order assigned to %s\n", selected.Name)
	return order, nil
}

func (s *FoodOrderingSystem) UpdateOrderStatus(orderId string, status enums.OrderStatus) error {
	order, exists := s.Orders[orderId]
	if !exists {
		return fmt.Errorf("order mathematically unmapped natively")
	}

	if order.Status == enums.Completed {
		return fmt.Errorf("order inherently tracked as securely completed")
	}

	if status == enums.Completed {
		order.Status = status
		s.Restaurants[order.RestaurantName].MarkOrderCompleted()
	}

	return nil
}
